elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/AutoGrowingPagedBuffer.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.elasticsearch.util.impl;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.CoderResult;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.LinkedList;
+
+import org.apache.http.nio.ContentEncoder;
+
+/**
+ * An automatically growing, paged buffer.
+ * <p>
+ * To be used when you need an infinite-size buffer,
+ * for example when your class is pushed content to
+ * and generate bytes from this content,
+ * but cannot always flush them to your consumer.
+ *
+ * @author Yoann Rodiere
+ */
+class AutoGrowingPagedBuffer {
+
+	/**
+	 * Size of pages allocated by this object.
+	 */
+	private final int pageSize;
+
+	/**
+	 * Filled pages to be written, in write order.
+	 */
+	private final Deque<ByteBuffer> needWritingPages = new LinkedList<>();
```

Comment:
I hate `LinkedList` in performance-sensitive code :)

I understand what you're thinking: we expect this to be very small and possibly not ever grow at.. still an ArrayList is _always_ more efficient, even a small one. There even has been a proposal to deprecate `LinkedList` as there's no reason to ever use it, but it was decided to keep it as it's actually interesting for students.

Reply:
in conclusion: a small ArrayList initialized at for example `5` should do the trick and still be somewhat better.

A point to remember is that while the size of allocations matter, the amount of objects being allocated matters as well, and finally the physical layout in memory; in both cases you have just pointers going to some other random location (wherever the page is) but with the array at least you load all pointers contiguously.

Reply:
Ok, thanks, I fixed this. I need a `Deque` though, so I'll go with an `ArrayDeque`.